crates/s3/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+    pub filename_prefix: String,
+}
+
+pub struct Uploader {
+    bucket: String,
+    filename_prefix: String,
+    client: Client,
+}
+
+impl Uploader {
+    pub async fn new(config: Config) -> Self {
+        let uploader = Self {
+            bucket: config.bucket,
+            filename_prefix: config.filename_prefix,
+            client: Client::new(&aws_config::load_from_env().await),
+        };
+        uploader.assert_credentials_are_usable().await;
+        uploader
+    }
+
+    /// Upload the bytes json encoded to the configured S3 bucket. Returns the
+    /// key under which the file can be queried
+    pub async fn upload(&self, id: String, content: impl Serialize) -> Result<String> {
+        let bytes = serde_json::to_vec(&content)?;
+        let encoded = self.gzip(&bytes)?;
+        let key = std::path::Path::new(&self.filename_prefix)
+            .join(format!("{id}.json"))
+            .to_str()
+            .context(anyhow!("invalid path: {id}"))?
+            .to_string();
+        self.client
+            .put_object()
+            .bucket(self.bucket.clone())
+            .key(key.clone())
+            .body(ByteStream::new(encoded.into()))
+            .content_encoding("gzip")
+            .content_type("application/json")
+            .send()
+            .await?;
+        Ok(key)
+    }
+
+    /// Uploads a small test file to verify that the credentials loaded from the
+    /// environment allow uploads to S3.
+    async fn assert_credentials_are_usable(&self) {
+        const DOCS_URL: &str = "https://docs.rs/aws-config/latest/aws_config/default_provider/credentials/struct.DefaultCredentialsChain.html";
+        self.upload(
+            "test".into(),
+            "test file to verify uploading capabilities".as_bytes(),
+        )
+        .await
+        .unwrap_or_else(|err| {
+            panic!(
+                "Could not upload test file to S3.\n Either disable uploads to S3 by removing the \
+                 s3_instance_upload_* arguments.\n Or make sure your environment variables are \
+                 set up to contain the correct AWS credentials.\n See {DOCS_URL} for more details \
+                 on that. \n{err:?}"
+            )
+        });
+        ()
+    }
+
+    /// Compresses the input bytes using Gzip.
+    fn gzip(&self, bytes: &[u8]) -> Result<Vec<u8>> {
+        let mut encoder = GzEncoder::new(bytes, Compression::best());
+        let mut encoded: Vec<u8> = Vec::with_capacity(bytes.len());
+        encoder.read_to_end(&mut encoded).context("gzip encoding")?;
+        Ok(encoded)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {super::*, flate2::read::GzDecoder, serde_json::json};
+
+    // This test requires AWS credentials to be set via env variables.
+    // See https://docs.rs/aws-config/latest/aws_config/default_provider/credentials/struct.DefaultCredentialsChain.html
+    // to know which arguments are expected and in what precedence they
+    // get loaded.
```

Comment:
How about configuring everything in the existing`docker-compose.yaml`?

Reply:
Even for the test, bucket and AWS credentials are secrets so I don't think we should expose them in the docker compose file. Or am I missing something?